docs/boring-bomb-architecture-doc.md

```diff
@@ -0,0 +1,115 @@
+    1. In the future, we’ll need to write a frontend for allowing verified media figures to create their own `DatasourceOwner` and attach is to 1 or more `Datasource`s, and we’ll need a way to verify they are indeed the owner of that `Datasource`. We will not do this now, and instead use manual updates to the DB until manual updates become too burdensome.
+    2. In the future if needed, we can add Auth functionality to the DB and write a frontend for Trust Assembly devs who don’t have access to the DB to add more `Datasources`, but in this MVP we’re making the decision to rely on direct DB inserts to add/update/delete `Datasources` and `DatasourceOwners`
+2. A cron job (or any function called periodically) will call a function on the Workers to fetch the `Datasource`s, and use them to insert jobs into the Worker Job Queue. Each job consists of a `Datasource` (for example a url like https://extelligence.substack.com/archive)
+3. The same Workers will subscribe to the job queue and one by one consume a job by using the `Datasources` to determine which URLs to fetch and parse. There will be some non-trivial logic here for how to use a `Datasource` along with existing `Articles` to determine which publications (i.e. specific article URL’s, like https://extelligence.substack.com/p/the-norwegian-butter-knife) need to have their headlines analyzed, and which already have `Article` s in the DB from a previous successfully completed job.
+4. Once the headline data is parsed, Workers will use OpenAI to analyze their original headlines & original article text to detect bias, and if the headline is biased then the Worker will use OpenAI to generate an unbiased headline.
+5. The worker will insert the unbiased headline into the DB as an `Article`. The Worker will also extract and insert the main article text into the `Article` . This isn’t useful for the MVP for Boring Bomb, but will be useful when we later move beyond the MVP to using a wiki to host main article text edits and Trust Assembly Adjudication.
+6. Finally, a user with the Trust Assembly browser extension visits a website, and the extension uses the API to fetch a headline for the article they’re reading. If no `Article` exists in our DB, then nothing will be changed about their website’s article. If one does exist, then the extension will use the unbiased headline from the `Article`
+
+## 🧑‍💼 Business Objects and Their Relations
+
+1. `Datasource`
+    1. id
+    2. parent_id (if the url to the same article has changed over time, then the new `Datasource` will have the prior `Datasource` as its parent_id)
```

Comment:
How worried are we that an article will change URLs over time? Does this happen a lot? Do you have any ideas on how we could handle this? Is this a beyond MVP concern?

Reply:
I think it happens very infrequently (my guess would be to 1% of articles), and when it does happen it's often because of some company-wide change to the structure of the URL path, such as when the company's backend team changes from v1 to v2 and stops responding to v1 routes. But I really don't know.

This is definitely not an MVP concern. I'm happy to remove it from the doc, and will do so.